[analyzer] New optin.taint.TaintedAlloc checker for catching unbounded memory allocation calls

dkrupp opened the pull request:
A new optional checker (optin.taint.TaintedAlloc) will warn if a memory allocation function (malloc, calloc, realloc, alloca, operator new[]) is called with a tainted (attacker controlled) size parameter.
A large, maliciously set size value can trigger memory exhaustion. To get this warning, the alpha.security.taint.TaintPropagation checker also needs to be switched on.

The warning will only be emitted, if the analyzer cannot prove that the size is below reasonable bounds (<SIZE_MAX/4).

There were no new reports with this extension on the following open source projects.

| Project | New Reports | Resolved Reports |
|---------|-------------|------------------|
| memcached | No reports | No reports 
| tmux | No reports | No reports 
| curl | No reports | No reports 
| twin | No reports | No reports 
| vim | No reports | No reports 
| openssl | No reports | No reports 
| sqlite | No reports | No reports 
| ffmpeg | No reports | No reports 
| postgres | No reports | No reports 
| xerces | No reports | No reports

dkrupp committed:
[analyzer] Adding taint analysis capability to unix.Malloc checker

unix.Malloc checker will warn if a memory allocation function
(malloc, calloc, realloc, alloca) is called with a tainted
(attacker controlled) size parameter.
A large, maliciously set size value can trigger memory exhaustion.
To get this warning, the alpha.security.taint.TaintPropagation checker
also needs to be switched on.

The warning will only be emitted, if the analyzer cannot prove
that the size is below reasonable bounds (<SIZE_MAX/4).

haoNoQ commented:
I think there should be a way to enable/disable this check separately because memory exhaustion / denial of service isn't necessarily something you care about when you enable taint analysis.

It's essential for web servers when the attacker is interested in interrupting their operation but not necessarily for personal devices where the attacker is interested only in gaining control of the device. For these applications it's more important to catch cases where malloc size and index used for access are coming from "different sources", eg. one is tainted and another isn't, doesn't matter which one.

For the same reason, I think the error node needs to be non-fatal. If you make it fatal, you lose the opportunity to catch this other type of bugs on the same path, which are 50% likely to be found on the same path, and are arguably much more severe.

Of course when your new check is disabled there won't be a sink so we'll still catch the other bug. But we don't actually want it to work this way; we'd rather have the user control the warnings they want to see with simple on-off flags without weird interactions between those flags.

So sink generation should be mostly a thing that "modeling" checkers do, when they're confident that we simply cannot continue analysis further. In this case I don't really see any problems with continuing the analysis so it should probably be a non-fatal error.

NagyDonat commented:
> I think the error node needs to be non-fatal.
Good point, I completely agree.

> For these applications it's more important to catch cases where malloc size and index used for access are coming from "different sources", eg. one is tainted and another isn't, doesn't matter which one.
I agree that we should enable the "pessimistic" handling (= report an error if we can't prove that the index is in bounds) for cases when the size is tainted, and I'm planning to create a commit that does this in `ArrayBoundCheckerV2`. (Previously passing a tainted value to `malloc` was an error, so this case simply never appeared.)

I would not emphasize "_different_ sources" in this context: a tainted index deserves careful handling even if the offset is also tainted. I see that this could produce false positives in situations like
```
char *f() {
  size_t n = get_tainted_value();
  if (!n)
    return NULL; // Rule out the case when n - 1 overflows.
  char *p = (char*)malloc(n);
  p[n-1] = '\0';
  return p;
}
```
where (AFAIK) the analyzer cannot conclude that `n - 1 < n` always holds under the known constraints. However, I think this wouldn't be very common, and if it turns out to be common, then we could improve the evaluation of the comparison operators by handling the case of `symbol + constant <=> same symbol`.

steakhal commented on clang/test/Analysis/malloc.c:
```diff
@@ -48,6 +49,45 @@ void myfoo(int *p);
+  size_t size;
+  int *p = malloc(sizeof(int)); 
+  scanf("%zu", &size);  
+  p = (int*) realloc((void*) p, size); // // expected-warning{{realloc is called with a tainted (potentially attacker controlled) value}}
```
```suggestion
  p = (int*) realloc((void*) p, size); // expected-warning{{realloc is called with a tainted (potentially attacker controlled) value}}
```

steakhal commented on clang/test/Analysis/malloc.c:
```diff
@@ -48,6 +49,45 @@ void myfoo(int *p);
+  size_t size;
+  int *p = alloca(sizeof(int)); 
+  scanf("%zu", &size);  
+  p = (int*) alloca(size); // // expected-warning{{alloca is called with a tainted (potentially attacker controlled) value}}  
```
```suggestion
  p = (int*) alloca(size); // expected-warning{{alloca is called with a tainted (potentially attacker controlled) value}}  
```

steakhal reviewed (commented):
The patch makes sense to me.
I'll not repeat the existing comments, they raise relevant concerns.

It would be nice to extend some test case with a tainted malloc to see how those note tags play out from the generic taint checker in this context. For this, I'd suggest you to have a look at some taint tests where we enable the `text` diagnostic output.

dkrupp committed:
Addressing review commnents

- Create a new optional checker optin.taint.TaintMalloc
- Add test case for testing taint diagnostic notes

dkrupp commented:
Thanks for the reviews. I updated the patch.

@haoNoQ 
- I changed the report to non-fatal
- I factored out the warning into a new checker optin.taint.TaintMalloc. This way the checker can be enabled separately. Of course, the alpha.security.taint.TaintPropagation checker is a prerequisite as indicated in the checker doc.

@steakhal 
- New test is added to the taint-diagnostic-visitor.c to test the taint related notes diagnostics.

-Minor changes addressed as requested.

Could you please check again?

steakhal reviewed (commented):
The patch makes sense to me.
Have you considered applying the same heuristic to C++ array new allocations?

I'll port this patch downstream to see how this would behave on the Juliet C++ benchmark or on some real-world code.

steakhal commented:
> I'll port this patch downstream to see how this would behave on the Juliet C++ benchmark or on some real-world code.

Ah nvm. llvm/main diverged quite a bit since 18.1.6. I can't just pick this one. Given this, I won't backport and test this PR.

NagyDonat reviewed (commented):
Oops, I noticed that I had some minor review commits that were left in a "Pending" state for at least a week. As the commit is not yet merged, I'm publishing them now.

dkrupp committed:
minor review fixes

NagyDonat commented on clang/include/clang/StaticAnalyzer/Checkers/Checkers.td:
```diff
@@ -1730,6 +1721,21 @@ def UnixAPIPortabilityChecker : Checker<"UnixAPI">,
+def TaintMallocChecker: Checker<"TaintMalloc">,
+  HelpText<"Check for memory allocations, where the size parameter "
+           "might be a tainted (attacker controlled) value.">,
+  Dependencies<[DynamicMemoryModeling]>,
```
Once the `TaintPropagation` checker will be brought out of alpha, it should be added as a dependency of this checker.

steakhal replied:
Why can't we make it a dependency now? (I don't have a strong opinion, I'm more just curious)

dkrupp replied:
I added the GenerictaintChecker as a dependency too. I think it makes sense to add it now so that we dont forget it later.

NagyDonat commented:
> Have you considered applying the same heuristic to C++ array new allocations?

Actually, I'm almost certain that this patch already affects the C++ array new allocations, because `MallocMemAux` is called from the callback which handles the array new operator.

@dkrupp Please add a C++ test file with a few testcases which show the behavior of your commit when an unconstrained tainted size is passed to `operator new[]`.

dkrupp commented:
> The patch makes sense to me. Have you considered applying the same heuristic to C++ array new allocations?
> 
> I'll port this patch downstream to see how this would behave on the Juliet C++ benchmark or on some real-world code.

I will check C++

> The patch makes sense to me. Have you considered applying the same heuristic to C++ array new allocations?
> 
> I'll port this patch downstream to see how this would behave on the Juliet C++ benchmark or on some real-world code.

I will consider adding the heuristic for C++ array new allocations in a follow-up patch.

dkrupp commented:
@NagyDonat , @steakhal  I fixed the additional remarks. 
Is there anything else that's need to be done before merging? Thanks.

dkrupp commented:
> @NagyDonat , @steakhal I fixed the additional remarks. Is there anything else that's need to be done before merging? Thanks.

I see now, that there is still one unaddressed remark from @NagyDonat regarding a new testcase for array new allocations. I will be adding it tomorrow...

dkrupp committed:
Fixing review comments

- Handling of C++ operator new[] allocation was added to the checker with test cases
- The checker is renamed to optin.taint.TaintAlloc, as besides malloc it handles the c++ new array allocations too
- Test cases and documentation was updated

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1779,18 +1797,76 @@ ProgramStateRef MallocChecker::MallocMemAux(CheckerContext &C,
+                                   llvm::ArrayRef<SymbolRef> TaintedSyms,
+                                   AllocationFamily Family) const {
+
+  if (!ChecksEnabled[CK_TaintAllocChecker])
+    return;
```
Move this check to `CheckTaintedness` -- there is no reason to calculate taint information if it won't be reported.

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1779,18 +1797,76 @@ ProgramStateRef MallocChecker::MallocMemAux(CheckerContext &C,
+  }
+}
+
+void MallocChecker::CheckTaintedness(CheckerContext &C, const CallEvent &Call,
```
Rename this to `checkTaintedness` with a lowercase 'c' to follow the global coding guidelines. I know that MallocChecker has lots of functions whose name start with an uppercase letter, but I think it's better to introduce new functions with conforming names and eventually switch to the standard naming scheme when there is a refactoring that already touches many functions.

(This is how I standardized the variable names in ArrayBoundV2.)

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1779,18 +1797,76 @@ ProgramStateRef MallocChecker::MallocMemAux(CheckerContext &C,
+    return;
+  auto [StateTooLarge, StateNotTooLarge] = State->assume(*Cmp);
+  if (!StateTooLarge && StateNotTooLarge) {
+    // we can prove that size is not too large so ok.
```
```suggestion
    // We can prove that size is not too large, so there is no issue."
```

NagyDonat reviewed (approved):
I'd say that the commit is acceptable as it is now, but I added several inline comments for minor prettification issues.

dkrupp committed:
formatting related review fixes

NagyDonat reviewed (approved):
LGTM, thanks for the update!

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -938,6 +938,53 @@ optin.portability.UnixAPI
+.. code-block:: c
+
+  void vulnerable(void) {
+    size_t size;
```
```suggestion
    size_t size = 0;
```
Let's initialize the value to be similar to the other cases.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1695,6 +1707,12 @@ MallocChecker::processNewAllocation(const CXXAllocatorCall &Call,
   // existing binding.
   SVal Target = Call.getObjectUnderConstruction();
+  if (Call.getOriginExpr()->isArray()) {
+    std::optional<const Expr *> SizeEx = NE->getArraySize();
+    if (SizeEx)
```
```suggestion
    if (auto SizeEx = NE->getArraySize())
```

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -938,6 +938,53 @@ optin.portability.UnixAPI
 
+.. _optin-taint-TaintAlloc:
+
+optin.taint.TaintAlloc (C, C++)
```
I was thinking of suggesting `TaintedAlloc` (same in past tense), as it sounds better to me. I have no strong opinion on this though.

steakhal reviewed (approved):
I join Donát, and I agree that this looks good as it is.
I had a handful of final remarks but I have no strong opinion on any of the raised points.
Merge this, once you considered them and took action if you agreed.

dkrupp committed:
Adding a fixme for the checker dependency, minor review fixes

dkrupp commented:
In the latest commit I fixed all remaining review comments.

GenericTaintchecker should be a dependency as mentioned in the FIXME, but it cannot be one until the checker is not a modeling checker. This separation will be done in a later follow-up patch. Until then, the documentation indicates the that alpha.security.taint.TaintPropagation checker should be switched on for this checker to work.

NagyDonat reviewed (approved):
The change LGTM. I agree with @steakhal that `TaintedAlloc` would be a slightly better name, but the current one is also acceptable.

dkrupp committed:
Renaming the checker to optin.taint.TaintedAlloc

dkrupp commented:
Now the checker is renamed to optin.taint.TaintedAlloc as requested by the reviewers.

steakhal commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -938,6 +938,53 @@ optin.portability.UnixAPI
+.. _optin-taint-TaintedAlloc:
+
+optin.taint.TaintedAlloc (C, C++)
+"""""""""""""""""""""""""""""""
```
```suggestion
"""""""""""""""""""""""""""""""""
```

dkrupp committed:
minor doc fix
